### ISSUE TITLE
AccountsDB returns Root data for Max commitment requests

### ISSUE THREAD

**Issue by CriesofCarrots** [solana-labs/solana#11161]
#### Problem
AccountsDB only retains account state back to the node's highest root (https://github.com/solana-labs/solana/blob/3fd16cea34fe18b751cbfe9a5f5520da21063fec/runtime/src/accounts_index.rs#L194-L205). This means that the correct account data may no longer be available for `max` commitment RPC requests, as the `highest_confirmed_root` is typically behind the node root.
In this case, AccountsDB returns data from the node root instead. AccountsDB does in fact report the newer slot that the data is from, but we ignore it here (https://github.com/solana-labs/solana/blob/f6e26f6c8c6276d84562047e4e4b43c6f9300010/runtime/src/bank.rs#L1822) and happily/erroneously report the `highest_confirmed_root` as the context slot in the rpc response.

#### Proposed Solution
Retain AccountsDB data until the node's observed `highest_confirmed_root`.
Also potentially do something with that slot in `Bank::get_account` to check that account data is from the slot we expect.

CC: @ryoqun @sakridge @mvines

**Comment by mvines** [comment]
Oh nice one!

**Comment by sakridge** [comment]
What is the client doing with old root data though?

**Comment by CriesofCarrots** [comment]
Max commitment allows the client an extra level of confidence, because the data has been rooted by not just this node, but 2/3+ of the cluster. (cc @aeyakovenko , if we need to generally debate the utility of max commitment)
However, given that we've chosen to offer it, we should probably make sure it's returning correct data.

**Comment by jeffwashington** [comment]
@jeffwashington: I may be able to work on this.

**Comment by CriesofCarrots** [comment]
@jeffwashington , did you ever look at this one at all?

**Comment by jeffwashington** [comment]
No. I have never done anything on this. Do you have any guidance on priority?

**Comment by CriesofCarrots** [comment]
> No. I have never done anything on this. Do you have any guidance on priority?

That's fine, I was just curious. I'd say it's still low-ish priority.
Not sure how straightforward the fix will end up being, but maybe it's something we can direct other contributors at, if the data inconsistency is problematic.
